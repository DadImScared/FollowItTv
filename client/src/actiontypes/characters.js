
export const ADD_CHARACTERS = 'characters/ADD_CHARACTERS';
