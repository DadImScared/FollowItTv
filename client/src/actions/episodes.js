
import { ADD_EPISODES } from '../actiontypes/episodes';

export const addEpisodes = (episodes) => {
  return {
    type: ADD_EPISODES,
    episodes
  };
};
