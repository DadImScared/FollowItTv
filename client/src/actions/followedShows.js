
import _ from 'lodash';
import axios from 'axios';

import axiosOptions from './axiosOptions';
import * as followedActionTypes from '../actiontypes/followedShows';
import { addShows } from './shows';

export const addDay = (day, shows) => {
  return {
    type: followedActionTypes.ADD_DAY,
    day,
    shows
  };
};

export const followShow = (days, id) => {
  return {
    type: followedActionTypes.FOLLOW_SHOW,
    days,
    id
  };
};

export const unfollowShow = (days, id) => {
  return {
    type: followedActionTypes.UNFOLLOW_SHOW,
    days,
    id
  };
};

export const postFollow = async (showId) => {
  return await axios.post(`/api/v1/followshow/${showId}/`, {}, axiosOptions());
};

export const addShowIds = (showIdList) => {
  return {
    type: followedActionTypes.ADD_SHOW_IDS,
    showIdList
  };
};

export const extractShowData = (data) => {
  let isArrayOfNumbers;
  const showIds = new Set();
  const showData = {};
  if (typeof data[0] === 'number') {
    isArrayOfNumbers = true;
  }
  else {
    data.forEach((show) => {
      showIds.add(show.id);
      showData[show.id] = show;
    });
  }
  return {
    showData,
    showIds: isArrayOfNumbers ? data:[...showIds]
  };
};

export const getFollowedShows = async (dispatch, day = '') => {
  try {
    const { data } = await axios.get(`/api/v1/followshows/${day ? `${_.capitalize(day)}/`:''}`, axiosOptions());
    const { showData, showIds } = extractShowData(data);
    if (day) {
      dispatch(addShows(showData));
    }
    dispatch(addShowIds(showIds));
  }
  catch (e) {
    throw e;
  }
};
