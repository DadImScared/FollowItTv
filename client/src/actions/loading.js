
import { LOADING_START, LOADING_END } from '../actiontypes/loading';

export const loadingStart = (requestId) => {
  return {
    type: LOADING_START,
    requestId
  };
};

export const loadingEnd = (requestId) => {
  return {
    type: LOADING_END,
    requestId
  };
};
