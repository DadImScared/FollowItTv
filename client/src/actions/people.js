
import { ADD_PEOPLE } from '../actiontypes/people';

export const addPeople = (people) => {
  return {
    type: ADD_PEOPLE,
    people
  };
};
