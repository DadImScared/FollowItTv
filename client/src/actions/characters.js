
import { ADD_CHARACTERS } from '../actiontypes/characters';
import { loadingEnd, loadingStart } from './loading';
import { castSchema } from '../schemas/person';
import { normalize } from 'normalizr';
import { baseUrl } from './seasons';
import axios from 'axios';

export const addCharacters = (entities) => ({
  type: ADD_CHARACTERS,
  ...entities
});

const getCast = async (showId) => {
  return await axios.get(`${baseUrl}shows/${showId}/cast`);
};

export const handleCastData = (data) => {
  const { entities: { person, character }, result } = normalize(data, castSchema);
  const characterIds = result.reduce((acc, { character }) => {
    const newAcc = [...acc];
    if (!newAcc.includes(character)) {
      newAcc.push(character);
    }
    return newAcc;
  }, []);
  return {
    person,
    character,
    result,
    characterIds
  };
};

export const requestCast = (showId) => {
  return async (dispatch) => {
    const loadingId = `GET_CAST_${showId}`;
    dispatch(loadingStart(loadingId));
    try {
      const { data } = await getCast(showId);
      const { person: people, character: characters, characterIds } = handleCastData(data);
      dispatch(addCharacters({ id: showId, people, characters, characterIds }));
      dispatch(loadingEnd(loadingId));
    }
    catch (e) {
      dispatch(loadingEnd(loadingId));
      throw e;
    }
  };
};
