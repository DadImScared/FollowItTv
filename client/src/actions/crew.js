
import axios from 'axios';
import _ from 'lodash';
import { normalize } from 'normalizr';

import { ADD_CREW } from '../actiontypes/crew';
import { loadingStart, loadingEnd } from './loading';
import { baseUrl } from './seasons';
import { crewSchema } from '../schemas/person';

export const addCrew = (entities) => ({
  type: ADD_CREW,
  ...entities
});

const getCrew = async (showId) => await axios.get(`${baseUrl}shows/${showId}/crew`);

export const handleCrewData = (data) => {
  const { entities, result } = normalize(data, crewSchema);
  const crewMembers = result.reduce((acc, crewMember) => {
    return {
      ...acc,
      [crewMember.person]: _.union(_.get(acc, crewMember.person, []), [ crewMember.type ])
    };
  }, {});
  return {
    ...entities,
    result,
    crewMembers
  };
};

export const requestCrew = (showId) => {
  return async (dispatch) => {
    const requestId = `GET_CREW_${showId}`;
    dispatch(loadingStart(requestId));
    try {
      const { data } = await getCrew(showId);
      const { person: people, crewMembers } = handleCrewData(data);
      dispatch(addCrew({ id: showId, people, crewMembers }));
      dispatch(loadingEnd(requestId));
    }
    catch(e) {
      dispatch(loadingEnd(requestId));
      throw e;
    }
  };
};
