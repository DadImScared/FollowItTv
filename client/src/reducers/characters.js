
import { ADD_CHARACTERS } from '../actiontypes/characters';

function characters(state = {}, action) {
  switch(action.type) {
    case ADD_CHARACTERS:
      return {
        ...state,
        ...action.characters
      };
    default:
      return state;
  }
}

export default characters;
