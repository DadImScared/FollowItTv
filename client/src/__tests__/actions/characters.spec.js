
import axios from 'axios';
import { addCharacters, requestCast, handleCastData } from '../../actions/characters';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';

jest.mock('axios');

const middlewares = [ thunk ];
const mockStore = configureMockStore(middlewares);

describe('characters', () => {
  test('addCharacters', () => {
    expect(addCharacters({
      characters: { 1: { id: 1, name: 'character1' } }
    })).toMatchSnapshot();
  });

  test('requestCast', async () => {
    axios.get.mockImplementationOnce(() => ({ data: [] }));
    const store = mockStore({ schedule: {} });
    await store.dispatch(requestCast(1));
    expect(store.getActions()).toMatchSnapshot();
  });

  test('handleCastData', () => {
    const data = [
      {
        person: { id: 5, name: 'person5' },
        character: { id: 1, name: 'character1' }
      },
      {
        person: { id: 2, name: 'person2' },
        character: { id: 2, name: 'character2' }
      }
    ];
    const results = handleCastData(data);
    expect(results).toMatchSnapshot();
  });
});
