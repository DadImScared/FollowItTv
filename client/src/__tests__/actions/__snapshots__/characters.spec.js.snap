// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`characters addCharacters 1`] = `
Object {
  "characters": Object {
    "1": Object {
      "id": 1,
      "name": "character1",
    },
  },
  "type": "characters/ADD_CHARACTERS",
}
`;

exports[`characters handleCastData 1`] = `
Object {
  "character": Object {
    "1": Object {
      "id": 1,
      "name": "character1",
      "people": 5,
    },
    "2": Object {
      "id": 2,
      "name": "character2",
      "people": 2,
    },
  },
  "characterIds": Array [
    1,
    2,
  ],
  "people": Object {
    "2": Object {
      "id": 2,
      "name": "person2",
    },
    "5": Object {
      "id": 5,
      "name": "person5",
    },
  },
  "result": Array [
    Object {
      "character": 1,
      "people": 5,
    },
    Object {
      "character": 2,
      "people": 2,
    },
  ],
}
`;

exports[`characters requestCast 1`] = `
Array [
  Object {
    "requestId": "GET_CAST_1",
    "type": "loading/LOADING_START",
  },
  Object {
    "requestId": "GET_CAST_1",
    "type": "loading/LOADING_END",
  },
]
`;
